js/app/src/components/Audio/Audio.test.ts

```diff
@@ -0,0 +1,150 @@
+import { test, describe, assert, afterEach, vi } from "vitest";
+import { spy } from "tinyspy";
+import { cleanup, render } from "@gradio/tootils";
+import event from "@testing-library/user-event";
+import { setupi18n } from "../../i18n";
+
+import Audio from "./Audio.svelte";
+import type { LoadingStatus } from "../StatusTracker/types";
+
+const loading_status = {
+	eta: 0,
+	queue_position: 1,
+	queue_size: 1,
+	status: "complete" as LoadingStatus["status"],
+	scroll_to_output: false,
+	visible: true,
+	fn_index: 0,
+	show_progress: "full" as LoadingStatus["show_progress"]
+};
+
+describe("Audio", () => {
+	afterEach(() => cleanup());
+
+	test("renders provided value and label", async () => {
+		const { getByTestId, container, queryAllByText } = render(Audio, {
+			show_label: true,
+			loading_status,
+			mode: "dynamic",
+			value: {
+				name: "https://gradio-builds.s3.amazonaws.com/demo-files/audio_sample.wav",
+				data: null,
+				is_file: true
+			},
+			label: "Audio Component",
+			root: "foo",
+			root_url: null,
+			streaming: false,
+			pending: false,
+			name: "bar",
+			source: "upload"
+		});
+
+		assert.equal(
+			getByTestId("Audio Component-dynamic-audio").src,
+			"foo/file=https://gradio-builds.s3.amazonaws.com/demo-files/audio_sample.wav"
+		);
+		assert(queryAllByText("Audio Component").length, 1);
+	});
+
+	test("hides label", async () => {
+		const { queryAllByText } = render(Audio, {
+			show_label: false,
+			loading_status,
+			mode: "dynamic",
+			value: {
+				name: "https://gradio-builds.s3.amazonaws.com/demo-files/audio_sample.wav",
+				data: null,
+				is_file: true
+			},
+			label: "Audio Component",
+			root: "foo",
+			root_url: null,
+			streaming: false,
+			pending: false,
+			name: "bar",
+			source: "upload"
+		});
+
+		assert(queryAllByText("Audio Component").length, 0);
+	});
+
+	test("upload sets change event", async () => {
+		setupi18n();
+		const { container, component } = render(Audio, {
+			show_label: false,
+			loading_status,
+			value: null,
+			mode: "dynamic",
+			label: "audio",
+			root: "foo",
+			root_url: null,
+			streaming: false,
+			name: "bar",
+			source: "upload"
+		});
+
+		const mock = spy();
+		component.$on("change", mock);
+
+		const item = container.querySelectorAll("input")[0];
+		console.log(item);
+		const file = new File(["hello"], "my-audio.wav", { type: "audio/wav" });
+		event.upload(item, file);
+		assert.equal(mock.callCount, 1);
```

Comment:
@pngwn @hannahblair @dawoodkhan82 why is the change event not being triggered here? Spent like an hour trying to figure it out 🥲

Reply:
The reaosn isn't isn't triggered is because the process of uploading a file is asynchronous.

We await `blobToBase64` here: https://github.com/gradio-app/gradio/blob/main/js/upload/src/Upload.svelte#L48
And you can see here it returns a Promise: https://github.com/gradio-app/gradio/blob/main/js/upload/src/utils.ts#L54-L62

There are two ways to solve this, one is to provide a  mock implementation the `Filereaders` `onload` event, so you can wait for it, but this is pretty messy and is also an implementation detail really (we may change to uploading the file later).

The second way feels a little circular but it would give the correct outcome. We can simply wait for the `change` event to fire, because that is the only real indication we have that the data is ready. It is a little circular because waiting for the `change` event to fire when we are testing that the change event fires is weird but it still allows us to test the functionality. We can add utility function like this:

```js 
async function wait_for_event(component, event) {
	return new Promise((r) => {
		component.$on(event, r);
	});
}
```

Then in the tests, just before we assert and after we call `event.upload(...)` we can do this:

```js
await wait_for_event(component, 'change');
```

The other error you are getting is weird, i have a fix but I'll take care of that in a different PR. Not sure why it is happening but i just don't think the test environment is really up to the job. I'll switch that over and when you merge in you should be okay.

Hope this helps! lmk if you have any more questions or want to pair.

Reply:
The changes i mentioned are in main now, so merging main into this branch + the above should help.